### PR TITLE
Feature/622/temporal coupling scm parser

### LINKED ISSUE #622: SCMLogParser: Calculate temporal coupling
# Feature request

[CodeMaat](https://github.com/adamtornhill/code-maat) calculates a very useful metric called temporal coupling. Unfortunately most development work now flows into the commercial version [CodeScene](https://codescene.io) and maat received its last code change years ago.

## Description

Calculate temporal coupling in CC. Temporal coupling is defined by Adam Tornhill in [Your Code as Crime Scene](https://www.amazon.com/Your-Code-Crime-Scene-Bottlenecks-ebook/dp/B00ZB5XWBI/) or the follow-up [Software Design X-Rays](https://www.amazon.com/Software-Design-X-Rays-Technical-Behavioral-ebook/dp/B07BVRLZ87/). 

Here's how I understood it: 

* Look at all instances, which can be commits/close time-windows/branches
* When file A is changed, how often is file B also changed? coupling A --> B
* Is the reverse also true or is file B changed without file A changing?
* given all the instances, how often is file A changed and file B as well? In percent, that is the temporal coupling between A --> B.

Please invest time to research this and compare it to the result calculated by codemaat for a couple of open source code bases.

### PR DESCRIPTION
**Pull request by alexhunziker**
# Edges for Temoral Coupling Returned by SCMLogParser

closes #622 

## Description

- SCMLogParser is now able to store edges to projects
- HighlyCoupledFiles Metric has been updated to also return the edges

Here is the map generated for CC with the new edges:
grafik

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFile.kt [hunk 1]
```diff
@@ -62,6 +63,26 @@ class VersionControlledFile internal constructor(
         return "$actualFilename with metrics $metricsMap"
     }
 
+    fun getEdgeList(): List<Edge> {
+        val edgeList = mutableListOf<Edge>()
+        metrics.flatMap { it.getEdges() }
+                .forEach { edge ->
+                    addEdgeToEdgeList(edge, edgeList)
+
```
Comment (NearW) [inline analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFile.kt]: remove empty line

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFile.kt [hunk 2]
```diff
@@ -62,6 +63,26 @@ class VersionControlledFile internal constructor(
...
+                }
+        return edgeList
+    }
+
+    private fun addEdgeToEdgeList(edge: Edge, edgeList: MutableList<Edge>) {
+        edgeList.forEach {
+            if (it.toNodeName == edge.toNodeName) {
+                it.attributes.toMutableMap().putAll(edge.attributes)
+                return@forEach
```
Comment (alexhunziker) [inline analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFile.kt]: This is probably wrong, we want to return nur return@forEach

**Comment by NearW** [comment]
@fatihi Should the metric on the edge be named `pairing-rate` as it is in codemaat or should we call it `temporal coupling`?

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/SCMLogParser.kt [hunk 3]
```diff
@@ -73,7 +73,8 @@ class SCMLogParser(private val input: InputStream = System.`in`,
                     "number_of_renames",
                     "range_of_weeks_with_commits",
                     "successive_weeks_of_commits",
-                    "weeks_with_commits"
+                    "weeks_with_commits",
+                    "highly_coupled_files"
```
Comment (NearW) [inline analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/SCMLogParser.kt]: why don't we need `medium_coupled_files` here as well?

**Comment by NearW** [comment]
@alexhunziker Please update the CHANGELOG.md and README.md accordingly

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/SCMLogParser.kt (hunk 3 above)
Comment (alexhunziker) [reply analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/SCMLogParser.kt]: *median_coupled_files ;) yes you are right, we can include this one here too. It's not strictly within the scope of the issue but as a boy-scout, it surely makes sense to calculate all metrics we can.

**Comment by fatihi** [comment]
@NearW Since we have another caluclation, I would prefer temporal coupling. Thanks for the hint

**Pull request merged**